application/comit_node/src/connection_pool.rs

```diff
@@ -0,0 +1,286 @@
+    fn new_outgoing_socket<B: BobSpawner>(
+        node_id: NodeId,
+        socket: Result<TcpStream, io::Error>,
+        connection_handle: Arc<Mutex<ConnectionState>>,
+        bob_spawner: B,
+    ) -> Result<BamClient, io::Error> {
+        let mut connection_state = connection_handle.lock().unwrap();
+        match *connection_state {
+            ConnectionState::Connected { ref client } => {
+                // We're already connected by the time we managed to connect.
+                // Forget about this new one and just use the old one.
+                match socket {
+                    Ok(_socket) => {
+                        debug!("Successfully connected to {} but we already have an exsting connection", node_id);
+                        // let the socket go out of scope and get dropped
+                        Ok(Arc::clone(&client))
+                    }
+                    Err(e) => {
+                        error!("Failed to connect to {} (but we already have an existing connection): {:?} ",node_id, e);
+                        Ok(Arc::clone(&client))
+                    }
+                }
+            }
+            // The most usual case: We were in Connecting and now we have a socket
+            // we want to transition to Connected
+            ConnectionState::Connecting { .. } => match socket {
+                Ok(socket) => Ok(Self::transition_to_connected(
+                    node_id,
+                    &mut *connection_state,
+                    Arc::clone(&connection_handle),
+                    bob_spawner,
+                    socket,
+                )),
+                Err(e) => {
+                    error!("Failed to connect to {}: {:?}", node_id, e);
+                    *connection_state = ConnectionState::Disconnected;
+                    Err(e)
+                }
+            },
+            // Somehow we got into Disconnected while connecting
+            // Just transition to Connected. (I guess?)
+            ConnectionState::Disconnected => match socket {
+                Ok(socket) => Ok(Self::transition_to_connected(
+                    node_id,
+                    &mut *connection_state,
+                    Arc::clone(&connection_handle),
+                    bob_spawner,
+                    socket,
+                )),
+                Err(e) => {
+                    error!("Failed to connect to {}: {:?}", node_id, e);
+                    Err(e)
+                }
+            },
+        }
+    }
+
+    /// Takes a server socket and updates the
+    pub fn new_incoming_socket<B: BobSpawner>(&self, socket: TcpStream, bob_spawner: B) {
+        let node_id = match socket.peer_addr() {
+            Ok(node_id) => node_id,
+            Err(e) => {
+                error!("Couldn't get peer address: {:?}", e);
+                return;
+            }
+        };
+        let mut connections = self.connections.write().unwrap();
+
+        let existing_connection = connections
```

Comment:
These lines are duplicated from above but I don't know if it is worth the effort to reduce the duplication.